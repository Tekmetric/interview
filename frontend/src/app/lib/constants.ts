export const API_BASE_URL = 'https://api.spacexdata.com/v4'

export const REVALIDATION_INTERVALS = {
  NEXT_LAUNCH: 60, // 1 minute
  LATEST_LAUNCH: 300, // 5 minutes
  UPCOMING_LAUNCHES: 300, // 5 minutes
  PAST_LAUNCHES: 900, // 15 minutes
  LAUNCH_STATS: 1800, // 30 minutes
  ROCKETS: 3600, // 1 hour
} as const

export const DEFAULT_PAGE_SIZE = 5

export const LAUNCH_SITES = [
  {
    id: 'ccafs',
    name: 'Cape Canaveral',
    latitude: 28.4922,
    longitude: -80.5217,
  },
  {
    id: 'vafb',
    name: 'Vandenberg',
    latitude: 34.632,
    longitude: -120.6107,
  },
  {
    id: 'stls',
    name: 'Starbase',
    latitude: 25.9971,
    longitude: -97.1554,
  },
] as const

export const MAP_CONFIG = {
  DEFAULT_CENTER: [35.8283, -98.5795],
  DEFAULT_ZOOM: 4,
  MAX_ZOOM: 19,
} as const
