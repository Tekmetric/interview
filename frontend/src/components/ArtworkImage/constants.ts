export const ADDITIONAL_INFO_CLOSE_BUTTON_SIZE_PX = 48
