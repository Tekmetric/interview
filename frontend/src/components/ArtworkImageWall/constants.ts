export const SINGLE_LIGHT_SOURCE_ARTWORK_ASPECT_RATIO_THRESHOLD = 0.75
