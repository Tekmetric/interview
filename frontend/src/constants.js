export const BASE_URL = 'https://fakestoreapi.com';
export const SORTING = {
  ASC: 'asc',
  DESC: 'desc',
}