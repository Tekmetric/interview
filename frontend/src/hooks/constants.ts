export const API_BASE_URL = process.env.REACT_APP_API_SERVER_URL || 'http://localhost:8080';
export const API_ENDPOINT = '/api/repair-services';
export const REPAIR_SERVICES_CACHE_KEY = 'repair-services';
