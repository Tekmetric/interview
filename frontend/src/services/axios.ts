import axios from 'axios';

const axiosClient = axios.create({
  baseURL:
    'https://vpic.nhtsa.dot.gov/api/vehicles',
    params: {
      format: 'json'
    }
});

export default axiosClient;
