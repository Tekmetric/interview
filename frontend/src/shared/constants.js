import dayjs from 'dayjs';

export const ROOT_URL = 'http://localhost:3001';

export const CARS = [
  {
    id: 1,
    brand: 'BMW',
    color: 'Black',
    description:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Phasellus venenatis ac nulla nec aliquam. Maecenas tincidunt metus non porta fringilla. Ut suscipit eu nibh et vehicula. Nulla ut velit sapien. Vestibulum sagittis semper nulla at congue. Nullam gravida, ipsum quis fermentum pretium, sem nisl interdum eros, non maximus lacus tortor.',
    model: '3 Series',
    engineCapacity: 1996,
    year: 2006,
    minPrice: 4000,
    maxPrice: 8000,
    url: 'https://www.topgear.com/sites/default/files/images/news-article/2015/07/75188e63e4d7cff5b227b81467219769/e90_1.jpg'
  },
  {
    id: 2,
    brand: 'Audi',
    color: 'White',
    description:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Phasellus venenatis ac nulla nec aliquam. Maecenas tincidunt metus non porta fringilla. Ut suscipit eu nibh et vehicula. Nulla ut velit sapien. Vestibulum sagittis semper nulla at congue. Nullam gravida, ipsum quis fermentum pretium, sem nisl interdum eros, non maximus lacus tortor.',
    model: 'A4',
    engineCapacity: 1993,
    year: 2015,
    minPrice: 15000,
    maxPrice: 25000,
    url: 'https://hips.hearstapps.com/hmg-prod/amv-prod-cad-assets/images/14q1/562749/2015-audi-a4-feature-car-and-driver-photo-581834-s-original.jpg?fill=2:1&resize=1200:*'
  },
  {
    id: 3,
    brand: 'Mercedes',
    color: 'Gray',
    description:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Phasellus venenatis ac nulla nec aliquam. Maecenas tincidunt metus non porta fringilla. Ut suscipit eu nibh et vehicula. Nulla ut velit sapien. Vestibulum sagittis semper nulla at congue. Nullam gravida, ipsum quis fermentum pretium, sem nisl interdum eros, non maximus lacus tortor.',
    model: 'C Class',
    engineCapacity: 1989,
    year: 2015,
    minPrice: 18000,
    maxPrice: 32000,
    url: 'https://www.carscoops.com/wp-content/uploads/2013/12/2015-Mercedes-C-Class-01.jpg'
  },
  {
    id: 4,
    brand: 'Kia',
    color: 'Penta Metal',
    description:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Phasellus venenatis ac nulla nec aliquam. Maecenas tincidunt metus non porta fringilla. Ut suscipit eu nibh et vehicula. Nulla ut velit sapien. Vestibulum sagittis semper nulla at congue. Nullam gravida, ipsum quis fermentum pretium, sem nisl interdum eros, non maximus lacus tortor.',
    model: 'Sportage',
    engineCapacity: 1588,
    year: 2022,
    minPrice: 25000,
    maxPrice: 49000,
    url: 'https://www.kia.ro/static/newsite/content/images/models//sportage-hev-2022/highlight/kia-sportage-2022-functionalitate-5.jpg'
  }
];

export const INITIAL_CAR_VALUES = {
  brand: '',
  color: '',
  description: '',
  model: '',
  engineCapacity: 1000,
  year: dayjs().year(),
  minPrice: 1000,
  maxPrice: 500000,
  url: ''
};
