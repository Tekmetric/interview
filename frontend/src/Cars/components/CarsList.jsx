import {
  Card,
  CardContent,
  Grid,
  CardMedia,
  Typography,
  CardActions,
  Button,
  Box,
  Pagination
} from '@mui/material';
import React, { useEffect, useState } from 'react';
import { CARS } from '../../shared/constants';
import { useNavigate, useSearchParams } from 'react-router-dom';
import { truncate } from '../../shared/helpers';
import CarsFilters from './CarsFilters';
//import logo from './logo.svg';

function CarsList() {
  const [cars, setCars] = useState(CARS);
  const navigate = useNavigate();
  const [page, setPage] = useState(1);
  const handlePageChange = (event, val) => {
    setPage(val);
  };
  const [searchParams] = useSearchParams();

  useEffect(() => {
    const newCars = CARS.filter(
      (car) =>
        car.brand.toLowerCase() === searchParams.get('brand') &&
        car.color.toLowerCase() === searchParams.get('color')
    );
    setCars(newCars);
  }, [searchParams]);

  return (
    <Box sx={{ position: 'relative', height: '100%' }}>
      <Box display="flex" flexDirection="column" alignItems="center" sx={{ mt: 5 }}>
        <CarsFilters />
        <Grid container spacing={2} sx={{ padding: '2rem 5rem' }}>
          {cars.map((car) => (
            <Grid item xs={12} md={4} lg={3} key={car.id}>
              <Card>
                <CardMedia sx={{ height: 140 }} image={car.url} title="green iguana" />
                <CardContent>
                  <Typography gutterBottom variant="h5" component="div">
                    {car.brand} - {car.model}
                  </Typography>
                  <Typography variant="body2" color="text.secondary">
                    {truncate(car.description, 180)}
                  </Typography>
                </CardContent>
                <CardActions>
                  <Button size="small" onClick={() => navigate(`${car.id}`)}>
                    Details
                  </Button>
                </CardActions>
              </Card>
            </Grid>
          ))}
        </Grid>
        <Pagination count={10} page={page} onChange={handlePageChange} sx={{ mt: 5 }} />
      </Box>
    </Box>
  );
}

export default CarsList;
