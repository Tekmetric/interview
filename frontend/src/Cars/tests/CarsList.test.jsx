import renderer from 'react-test-renderer';
import CarsList from '../components/CarsList';
import { BrowserRouter } from 'react-router-dom';
import { QueryClient, QueryClientProvider } from 'react-query';
import { CARS } from '../../shared/constants';
import { render, fireEvent } from '@testing-library/react';
import nock from 'nock';
import axios from 'axios';
import httpAdapter from 'axios/lib/adapters/http';

const queryClient = new QueryClient();
// eslint-disable-next-line react/prop-types
const wrapper = ({ children }) => (
  <BrowserRouter>
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  </BrowserRouter>
);

const host = 'http://localhost:3001';
axios.defaults.host = host;
axios.defaults.adapter = httpAdapter;

nock('http://localhost:3001')
  .get('/cars')
  .query({ page: 1 })
  .reply(200, { cars: CARS, totalPages: 2 });

nock('http://localhost:3001')
  .get('/cars')
  .query({ page: 1, brand: 'bmw', color: 'all' })
  .reply(200, { cars: CARS.filter((car) => car.brand.toLowerCase() === 'bmw'), totalPages: 2 });

nock('http://localhost:3001')
  .get('/cars/brands')
  .reply(200, [
    { key: 'all', value: 'All' },
    { key: 'audi', value: 'Audi' },
    { key: 'bmw', value: 'BMW' }
  ]);

nock('http://localhost:3001')
  .get('/cars/colors')
  .reply(200, [
    { key: 'all', value: 'All' },
    { key: 'black', value: 'Black' },
    { key: 'gray', value: 'Gray' }
  ]);

describe('rendering list of cars', () => {
  it('matches the snapshot', () => {
    const component = renderer.create(wrapper(<CarsList />));

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a list of cars', async () => {
    const result = render(<CarsList />, { wrapper });

    for (let car of CARS) {
      await result.findByText(`${car.brand} - ${car.model}`);

      expect(result.getByText(`${car.brand} - ${car.model}`)).toBeTruthy();
    }
  });

  it('renders a list of cars filtered by brand', async () => {
    const result = render(<CarsList />, { wrapper });

    await result.findByLabelText('Brand');

    const brandSelector = document.querySelector('input.MuiSelect-nativeInput');
    fireEvent.change(brandSelector, { target: { value: 'bmw' } });

    await result.findAllByText('BMW - 3 Series');
    expect(result.getAllByText('BMW - 3 Series').length).toEqual(1);
  });
});
