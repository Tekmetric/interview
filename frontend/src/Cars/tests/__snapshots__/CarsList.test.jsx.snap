// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering list of cars matches the snapshot 1`] = `null`;
